nexus/test-utils/src/background.rs

```diff
@@ -69,10 +69,45 @@ pub async fn activate_background_task(
             .execute_and_parse_unwrap::<BackgroundTask>()
             .await;
 
-            if matches!(&task.current, CurrentStatus::Idle)
-                && most_recent_start_time(&task) > last_start
-            {
-                Ok(task)
+            // Wait until the task has actually run and then is idle
+            if matches!(&task.current, CurrentStatus::Idle) {
+                let current_activate = most_recent_activate_time(&task);
+                match (current_activate, last_activate) {
+                    (None, None) => {
+                        // task is idle but it hasn't started yet, and it was
+                        // never previously activated
+                        Err(CondCheckError::<()>::NotYet)
+                    }
+
+                    (Some(_), None) => {
+                        // task was activated for the first time by this
+                        // function call, and it's done now (because the task is
+                        // idle)
+                        Ok(task)
+                    }
+
+                    (None, Some(_)) => {
+                        // the task is idle (due to the check above) but
+                        // `most_recent_activate_time` returned None, implying
+                        // that the LastResult is NeverCompleted? the Some in
+                        // the second part of the tuple means this ran before,
+                        // so panic here.
+                        panic!("task is idle, but there's no activate time?!");
```

Comment:
We panic here because this should not be possible, right?

Is there any chance the `task.current` could change between we `matches` on line 73 and make the call to `most_recent_activate_time()`?

Reply:
The state of the background task could change yeah but we'll pick that up when it gets re-polled.

The state of the world in that part of the `match` shouldn't ever be possible, no, I think it's appropriate to panic there.